### PR TITLE
Automation for running NVIDIA MLPerf SSD training

### PR DESCRIPTION
**Pull request by Maxusmusti**
Get nvidiadl ssd training benchmark working with new auto-updated image

NOTE: `test-path: recipes run_cocodataset_nvidiadl_ssd`

### CODE REVIEW

**Comment by kpouget** [comment]
```
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | roles/benchmarking_run_nvidiadl_ssd/tasks/main.yml:48
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | TASK: benchmarking_run_nvidiadl_ssd : Wait for the benchmark completion
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | ----- FAILED ----
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | 
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | <command> oc get pod/benchmarking-nvidiadl-ssd --no-headers -ocustom-columns=phase:status.phase -n default
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | 
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | <stdout> Running
2022-01-19 07:22:45,091 p=9334 u=psap-ci-runner n=ansible | ----- FAILED ----
2022-01-19 07:22:45,093 p=9334 u=psap-ci-runner n=ansible | PLAY RECAP *********************************************************************
```
@Maxusmusti , I see that the benchmark Pod logs are not saved, at least in case of a [failure](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/316/pull-ci-openshift-psap-ci-artifacts-master-test-commit/1483679124391202816/artifacts/test-commit/test-commit/artifacts/recipes_run_cocodataset_nvidiadl_ssd/012__benchmarking__run_nvidiadl_ssd/_ansible.log) as above, could you make sure it's always stored so that we know what happened in the execution?

**Comment by kpouget** [comment]
```
 {"component":"entrypoint","file":"prow/entrypoint/run.go:165","func":"k8s.io/test-infra/prow/entrypoint.Options.ExecuteProcess","level":"error","msg":"Process did not finish before 2h0m0s timeout","severity":"error","time":"2022-01-19T09:39:52Z"} 
```
cluster got terminated, we need to talk about that :) [=that was expected for me]

**Comment by kpouget** [comment]
Hey Mustafa,here is what I have in my scripts:

```
# prepare dataset according to download_dataset.sh

if [ ! -f ${DATASET_DIR}/annotations/bbox_only_instances_val2017.json ]; then
    echo "Prepare instances_val2017.json ..."
    ./prepare-json.py --keep-keys \
        "${DATASET_DIR}/annotations/instances_val2017.json" \
        "${DATASET_DIR}/annotations/bbox_only_instances_val2017.json"
fi

if [ ! -f ${DATASET_DIR}/annotations/bbox_only_instances_train2017.json ]; then
    echo "Prepare instances_train2017.json ..."
    ./prepare-json.py \
        "${DATASET_DIR}/annotations/instances_train2017.json" \
        "${DATASET_DIR}/annotations/bbox_only_instances_train2017.json"
fi
```

**Comment by Maxusmusti** [comment]
> if [ ! -f ${DATASET_DIR}/annotations/bbox_only_instances_train2017.json ]; then
>     echo "Prepare instances_train2017.json ..."
>     ./prepare-json.py \
>         "${DATASET_DIR}/annotations/instances_train2017.json" \
>         "${DATASET_DIR}/annotations/bbox_only_instances_train2017.json"
> fi

ah I see, thanks!

**Comment by Maxusmusti** [comment]
Got the ssd training benchmark working/running

**Comment by kpouget** [comment]
hum, not sure [yet](https://gcsweb-ciapps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/316/pull-ci-openshift-psap-ci-artifacts-master-test-commit/1486725458543251456/artifacts/test-commit/test-commit/artifacts/recipes_run_cocodataset_nvidiadl_ssd/012__benchmarking__run_nvidiadl_ssd/pod_run-nvidiadl-ssd.log) ;-)

```
:::MLLOG {"namespace": "", "time_ms": 1643303277049, "event_type": "INTERVAL_END", "key": "epoch_stop", "value": null, "metadata": {"file": "train.py", "lineno": 427, "epoch_num": 2}}
:::MLLOG {"namespace": "", "time_ms": 1643303277262, "event_type": "INTERVAL_END", "key": "run_stop", "value": null, "metadata": {"file": "train.py", "lineno": 449, "status": "aborted"}}
```

I had to monkey patch the [launcher](https://github.com/kpouget/matrix-benchmark/blob/hpc/exec/mlperf/my_run_and_time.sh#L197) so that errors are properly reported, when the `train.py` fails (without it, `bind_launch.py` never throws any error)
```
echo "Patching 'bind_launch.py' to err-exit on failure ..."
sed 's/process.wait()$/if process.wait(): sys.exit(1)/' -i bind_launch.py
```

[here](https://github.com/kpouget/matrix_benchmark_results/blob/dgxa100/dgx-benchmark/20211209/gpu%3Dfull_1gpu_x_1pod/run_20211213_2015.d988/run-ssd--1-tzxw4.log) is a log file of a complete execution

and I guess you should add 'set -x' in the script to show what's going on before the training is launched

File: roles/benchmarking_run_nvidiadl_ssd/files/entrypoint.sh [hunk 1]
```diff
@@ -1,10 +1,49 @@
 #! /bin/bash
```
Comment (kpouget) [inline roles/benchmarking_run_nvidiadl_ssd/files/entrypoint.sh:1]: can you add
```
set -o pipefail
set -o errexit
set -o nounset
set -x
```

File: roles/benchmarking_run_nvidiadl_ssd/files/entrypoint.sh [hunk 2]
```diff
@@ -1,10 +1,49 @@
...
-          --mode benchmark-training \
-          --benchmark-warmup 100 \
-          --benchmark-iterations 100 \
-          --data /storage
+export NCCL_DEBUG=INFO
+export CUDA_VISIBLE_DEVICES=0
+exec python -u -m bind_launch --nsockets_per_node=1 --ncores_per_socket=1 --nproc_per_node=1 \
+     train.py --epochs 2 \
+              --warmup-factor 0 \
+              --threshold=0.25 \
```
Comment (kpouget) [inline roles/benchmarking_run_nvidiadl_ssd/files/entrypoint.sh:45]: can you make the number of `epochs` and `threshold` configurable from the entrypoint?
by default, let it empty, and in the script, if empty, use these defaults values:
```
+ THRESHOLD=0.23
+ NUM_EPOCHS=80
```
> --threshold=0.25 
I think threshold `0.25` is unreachable (eg, asymptotical to it), `0.23` is the sensible limit ;-)

**Review by kpouget — CHANGES_REQUESTED** [review]
thanks @Maxusmusti, looks good
I added a few details to update

what happens with epochs = 1, is it running in half the time?

**Comment by Maxusmusti** [comment]
NOTE: Working on getting the configurable epochs/thresholds, should be done very soon

**Comment by Maxusmusti** [comment]
Testing timing of 3 epochs with a preset seed

**Comment by kpouget** [comment]
/label tide/merge-method-squash

I forgot activate that for https://github.com/openshift-psap/ci-artifacts/pull/292 so we merged 44 commits in the tree :# (default in this repo is `rebase`)

**Comment by kpouget** [comment]
@Maxusmusti do you know what the benchmark was doing in Pam's original script?
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/323/pull-ci-openshift-psap-ci-artifacts-master-test-commit/1488444549335355392/artifacts/test-commit/test-commit/artifacts/recipes_run_cocodataset_nvidiadl_ssd/012__benchmarking__run_nvidiadl_ssd/pod_run-nvidiadl-ssd.log

```
DLL 2022-02-01 10:31:35.406587 - () avg_img/sec : 32.33001815606221  med_img/sec : 32.315785352001804  min_img/sec : 31.573153473246816  max_img/sec : 32.95698772649411 
Done benchmarking. Total images: 3200	total time: 98.979	Average images/sec: 32.330	Median images/sec: 32.316
Training performance = 32.3157844543457 FPS
DLL 2022-02-01 10:31:35.597567 - (0,) time : 206.60495257377625 
WARNING:root:DALI iterator does not support resetting while epoch is not finished. Ignoring...
DLL 2022-02-01 10:31:35.597850 - () total time : 206.60495257377625 
DLL 2022-02-01 10:31:35.597890 - () 
```

I wonder why it terminates properly in 206 seconds?

**Review by kpouget — COMMENTED** [review]
arrh, sorry @Maxusmusti , I made those comments maybe yesterday morning, but I forgot to submit the review :/

**Comment by Maxusmusti** [comment]
> @Maxusmusti do you know what the benchmark was doing in Pam's original script? https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/323/pull-ci-openshift-psap-ci-artifacts-master-test-commit/1488444549335355392/artifacts/test-commit/test-commit/artifacts/recipes_run_cocodataset_nvidiadl_ssd/012__benchmarking__run_nvidiadl_ssd/pod_run-nvidiadl-ssd.log
> 
> ```
> DLL 2022-02-01 10:31:35.406587 - () avg_img/sec : 32.33001815606221  med_img/sec : 32.315785352001804  min_img/sec : 31.573153473246816  max_img/sec : 32.95698772649411 
> Done benchmarking. Total images: 3200	total time: 98.979	Average images/sec: 32.330	Median images/sec: 32.316
> Training performance = 32.3157844543457 FPS
> DLL 2022-02-01 10:31:35.597567 - (0,) time : 206.60495257377625 
> WARNING:root:DALI iterator does not support resetting while epoch is not finished. Ignoring...
> DLL 2022-02-01 10:31:35.597850 - () total time : 206.60495257377625 
> DLL 2022-02-01 10:31:35.597890 - () 
> ```
> 
> I wonder why it terminates properly in 206 seconds?

Rather than training for a certain number of epochs with a specified threshold, her script simply trains for 100 iterations (3200 total photos). The sample/sec rate actually appears to be 3x faster now then it was back when this was run.

File: roles/benchmarking_run_nvidiadl_ssd/tasks/main.yml [hunk 3]
```diff
@@ -45,33 +45,35 @@
   command:
     oc create -f "{{ artifact_extra_logs_dir }}/001_pod_run-nvidiadl-ssd.yml"
 
-- name: Wait for the benchmark completion
```
Comment (kpouget) [inline roles/benchmarking_run_nvidiadl_ssd/tasks/main.yml:48]: a bit above here (line 25), could you remove `-n {{ benchmarking_namespace }}` from `oc create` and move it below to `oc apply`?

will make it easier to reuse the files, as in 
```
oc apply -f https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/316/pull-ci-openshift-psap-ci-artifacts-master-test-commit/1488755158606155776/artifacts/test-commit/test-commit/artifacts/recipes_run_cocodataset_nvidiadl_ssd/012__benchmarking__run_nvidiadl_ssd/001_pod_run-nvidiadl-ssd.yml
```
currently the namespace must be `default`

and same idea for https://github.com/Maxusmusti/ci-artifacts/blob/cocobench/roles/benchmarking_run_nvidiadl_ssd/templates/nvidiadl-ssd-pod.yml.j2

if you remove `namespace: {{ benchmarking_namespace }}` from the file, and only define the namespace in `oc create -f "{{ artifact_extra_logs_dir }}/001_pod_run-nvidiadl-ssd.yml" -n "{{ benchmarking_namespace }}"`

**Comment by Maxusmusti** [comment]
Currently translating results from personal cluster testing + V100 testing to CI. Also currently in the process of uploading COCO dataset into S3 bucket (it's going incredibly slow...~150KB/s).

**Comment by kpouget** [comment]
@Maxusmusti is this PR ready for merging? I'll have another look tomorrow morning, but I guess we can merge it even if the hyperparameter values are not settled and the failure threshold is not defined

WDYT?

**Comment by Maxusmusti** [comment]
> @Maxusmusti is this PR ready for merging? I'll have another look tomorrow morning, but I guess we can merge it even if the hyperparameter values are not settled and the failure threshold is not defined
> 
> WDYT?

Yeah, this PR is ready for merge, as it does what it needs to in replacing the old benchmark. I'll test it one more time after rebasing on your recent changes, but otherwise good to go. I'll create a new PR for switching to S3 bucket and adding failure conditions 👍

**Comment by Maxusmusti** [comment]
Should be gtg

**Comment by kpouget** [comment]
thanks @Maxusmusti, seems good to me!

/approve
/lgtm

**Comment by kpouget** [comment]
> Not mergeable. Merge type "squash" disallowed by repo settings 

Tide cannot squash, seems to be caused by a [configuration issue](https://github.com/kubernetes/test-infra/issues/24141), but I cannot fixed it as they suggested :/ (Tide already has read/write access to the PRs)

"Squashing and merging" manually
